documentation/src/main/asciidoc/user_guide/query.adoc

```diff
@@ -92,73 +144,468 @@ for (Object book : objectList) {
+            .build();
+
+List<Book> list = q.list();
+
+----
+
+Finally, we can use an <<query.ickle,Ickle>> query directly, allowing for Lucene syntax in one or more predicates:
 
 [source,java]
 ----
+import org.infinispan.query.dsl.QueryFactory;
+import org.infinispan.query.dsl.Query;
 
-@Transformable(transformer = CustomTransformer.class)
-public class CustomKey {
-   ...
-}
+// get the query factory:
+QueryFactory queryFactory = Search.getQueryFactory(cache);
 
-public class CustomTransformer implements Transformer {
-   @Override
-   public Object fromString(String s) {
-      ...
-      return new CustomKey(...);
-   }
 
-   @Override
-   public String toString(Object customType) {
-      CustomKey ck = (CustomKey) customType;
-      return ...
-   }
-}
+Query q = queryFactory.create("from Book b where b.author.name = 'Stephen' and " +
+                "b.description : (+'dark' -'tower')");
+
+List<Book> list = q.list();
 
 ----
 
-===== Registering a Transformer programmatically
-Using this technique, you don't have to annotated your custom key type:
+==== Indexing
+
+Indexing in Infinispan happens on a per-cache basis and by default a cache is not indexed. Enabling indexing is not mandatory but queries using an index will
+have a vastly superior performance. On the other hand, enabling indexing can impact negatively the write throughput of a cluster, so make sure to check the <<query.performance, query performance guide>> for some strategies to minimize this impact depending on the cache type and use case.
+
+===== Configuration
+
+====== General format
+
+To enable indexing via XML, you need to add the `<indexing>` element plus the `index` (<<query.index-mode,index mode>>)
+to your cache configuration, and optionally pass additional properties.
+
+[source,xml]
+----
+<infinispan>
+   <cache-container default-cache="default">
+      <replicated-cache name="default">
+         <indexing index="ALL">
+            <property name="property.name">some value</property>
+         </indexing>
+      </replicated-cache>
+   </cache-container>
+</infinispan>
+
+----
+
+Programmatic:
 
 [source,java]
 ----
-org.infinispan.query.SearchManager.registerKeyTransformer(Class<?>, Class<? extends Transformer>)
+import org.infinispan.configuration.cache.*;
+
+ConfigurationBuilder cacheCfg = ...
+cacheCfg.indexing().index(Index.ALL)
+      .addProperty("property name", "propery value")
 ----
 
-=== Configuration [[query:configuration]]
-==== Configuration via XML [[query:configuration-xml]]
-To enable indexing via XML, you need to add the `<indexing />` element to your cache configuration, and optionally pass additional properties to the embedded Hibernate Search engine:
+====== Index names
+
+Each property inside the `index` element is prefixed with the index name, example:
+
+[source,xml]
+----
+    ...
+    <indexing index="ALL">
+        <property name="org.infinispan.sample.Car.directory_provider">ram</property>
```

Comment:
This section is presenting how index naming works; for the `Boat` entity, it is using a shorthand index name. The goal was to show the possible ways of configuring it.